### DISCUSSION TITLE
Is it possible to make two-way SSL asynchronous?

### DISCUSSION THREAD

**Discussion by msl12** [emqx/emqx#5831] [Q&A] [answered]
Is it possible to make two-way SSL asynchronous?

After the client verifies the server ssl certificate is correct, it can establish a connection successfully and the client can start to publish messages. This process is like **one-way SSL**.

Then the server would check the client ssl certificate asynchronously. And during the checking time, the client could publish messages normally, and the server could process these messages and response ACK. Once the server asynchronously find something wrong with the client ssl certificate, then the server would terminate the connection with client.

Can such this requirement be achieved? If possible, how should I do?

Thanks for your help!

------------------------------------------------------------------------------------------

有可能让SSL双向认证异步吗？

客户端验证完服务端的证书正确后，就可以建立好连接，并且可以开始上报消息。这时的过程就像**单向认证**。

然后服务端异步地去验证客户端的证书是否正确。异步检查期间，客户端可以正常上报消息并且服务端也可以处理并返回ACK。如果服务端异步检查发现客户端的证书不正确，就中断连接。

请问这个场景可以实现吗？该怎么做呢？

非常感谢！

**Comment by zmstone** [comment, marked as answer]
"one-way SSL" in practice is only to skip verifying the peer's certificate (or allow the peer not to send one),
maybe it can save some CPU resources in verifying the certificates,
but it does not make TLS handshake less expensive.

What you are requesting, is more like the zero-RTT (early data) from TLS 1.3.
Unfortunately Erlang's SSL lib does not support zero RTT.
